File: v2/java/part5/graph/Boj7576.java
package part5.graph;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Boj7576 {
    static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static final Deque<Triple> dq = new ArrayDeque<>();
    static int[][] box;
    static StringTokenizer st;
    static int N, M, ans, minus;

    static class Triple {
        int x, y, days;

        public Triple(int x, int y, int days) {
            this.x = x;
            this.y = y;
            this.days = days;
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        box = new int[M][N];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < N; j++) {
                int k = Integer.parseInt(st.nextToken());
                if (k == 1) dq.add(new Triple(i, j, 0));
                else if (k == -1) minus++;
                box[i][j] = k;
            }
        }

        if (dq.size() == N * M - minus) System.out.println(0);
        else if (solve()) System.out.println(ans);
        else System.out.println(-1);
    }

    static boolean solve() {
        final int[] dx = { -1, 1, 0, 0 };
        final int[] dy = { 0, 0, -1, 1 };

        while (!dq.isEmpty()) {
            Triple t = dq.removeFirst();
            ans = Math.max(t.days, ans);

            for (int i = 0; i < 4; i++) {
                int x = t.x + dx[i];
                int y = t.y + dy[i];

                if (x < 0 || y < 0 || x >= M || y >= N)
                    continue;

                if (box[x][y] == 0) {
                    box[x][y] = 1;
                    dq.add(new Triple(x, y, t.days + 1));
                }
            }
        }

        for (int i = 0; i < M; i++)
            for (int j = 0; j < N; j++)
                if (box[i][j] == 0)
                    return false;

        return true;
    }
}

File: Kotlin/src/main/kotlin/beginner/d_GraphSearch/boj_7576.kt
package beginner.d_GraphSearch

import java.util.StringTokenizer

private data class Point(val x: Int, val y: Int, val day: Int)

private var ans = 0
private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)
private val Pt = ArrayList<Pair<Int, Int>>()
private lateinit var Matrix: Array<IntArray>

private fun check_arr(): Boolean {
    for (mat in Matrix) for (m in mat) if (m == 0) return false
    return true
}

private fun solve() {
    val q = ArrayDeque<Point>()
    for (pr in Pt)
        q.add(Point(pr.first, pr.second, 0))

    while (q.isNotEmpty()) {
        val (nx, ny, nd) = q.removeFirst()
        ans = nd

        for (i in 0 until 4) {
            val tx = nx + dx[i]
            val ty = ny + dy[i]

            if (tx < 0 || ty < 0 || tx >= Matrix.size || ty >= Matrix[nx].size)
                continue

            if (Matrix[tx][ty] == 0) {
                Matrix[tx][ty] = 1
                q.add(Point(tx, ty, nd + 1))
            }
        }
    }
}

fun boj_7576() = with(System.`in`.bufferedReader()) {
    val (m, n) = readLine()
        .split(" ")
        .map { it.toInt() }

    Matrix = Array(n) { IntArray(m) }

    for (i in 0 until n) {
        val st = StringTokenizer(readLine())

        for (j in 0 until m) {
            val x = st.nextToken().toInt()
            if (x == 1) Pt.add(Pair(i, j))
            Matrix[i][j] = x
        }
    }

    solve()

    if (check_arr()) print(ans)
    else print(-1)
}